File: 05.py
#!/usr/bin/env python3
from sys import stdin
import re

def lines(xs):
    (a,b,c,d) = xs
    if   a==c: return [(a,x) for x in range(min(b,d),max(b,d)+1)]
    elif b==d: return [(x,b) for x in range(min(a,c),max(a,c)+1)]
    return []

def diags(xs):
    (a,b,c,d) = xs
    if (a==c or b==d): return []
    dx, dy = 1 if a < c else -1, 1 if b < d else -1
    l = max(abs(c - a), abs(d - b))
    return [(a + i*dx, b + i*dy) for i in range(l+1)]

ls    = (l.strip() for l in stdin.readlines())
ls    = [tuple(map(int,re.findall(r'\d+',l))) for l in ls]
ms    = [max(col) for col in zip(*ls)]
xm,ym = (max(ms[0],ms[2]), max(ms[1],ms[3]))
arr   = [[0] * (ym+1) for i in range(xm+1)]

for a,b in sum(map(lines, ls),[]): arr[b][a]+=1
p1    = sum(1 for _ in filter(lambda x: x>1, sum(arr,[])))

for a,b in sum(map(diags, ls),[]): arr[b][a]+=1
p2    = sum(1 for _ in filter(lambda x: x>1, sum(arr,[])))

print (f"Part 1: { p1 }, Part 2: { p2 }")

File: 05.scala
#!/usr/bin/env scala
import scala.io.Source.stdin

def lines(a:Int, b:Int, c:Int, d:Int) =
  if (a==c)      (List(b,d).min to List(b,d).max).map((a,_))
  else if (b==d) (List(a,c).min to List(a,c).max).map((_,b))
  else            List()    

def diags(a:Int, b:Int, c:Int, d:Int) =
  val (dx, dy) = (if a<c then 1 else -1, if b<d then 1 else -1)
  val l = List((c-a).abs, (d-b).abs).max
  if (a==c || b==d) List()
  else             (0 to l).map(i=>(a+i*dx, b+i*dy))

@main def aoc(): Unit =
  val ls   = stdin.getLines().map(("\\d+".r).findAllIn(_)
             .map(_.toInt).toList).map(x=>(x(0),x(1),x(2),x(3))).toList
  val mx   = ls.map(_.toList.max).max + 1
  val updt = (v:Vector[Int], x:(Int,Int))=>
             { val i = x._1*mx+x._2; v.updated(i, v(i)+1) }
  val p1   = ls.flatMap(lines.tupled).foldLeft(Vector.fill(mx*mx)(0))(updt)
  val p2   = ls.flatMap(diags.tupled).foldLeft(p1)(updt)

  println(s"Part 1: ${ p1.count(_>1) }, Part 2: ${ p2.count(_>1) }")
